tests/python/unittest/common.py

```diff
@@ -30,3 +35,190 @@ def assertRaises(expected_exception, func, *args, **kwargs):
     else:
         # Did not raise exception
         assert False, "%s did not raise %s" % (func.__name__, expected_exception.__name__)
+
+def default_logger():
+    """A logger used to output seed information to nosetests logs."""
+    logger = logging.getLogger(__name__)
+    # getLogger() lookups will return the same logger, but only add the handler once.
+    if not len(logger.handlers):
+        handler = logging.StreamHandler(sys.stderr)
+        handler.setFormatter(logging.Formatter('[%(levelname)s] %(message)s'))
+        logger.addHandler(handler)
+        if (logger.getEffectiveLevel() == logging.NOTSET):
+            logger.setLevel(logging.INFO)
+    return logger
+
+@contextmanager
+def random_seed(seed=None):
+    """
+    Runs a code block with a new seed for np, mx and python's random.
+
+    Parameters
+    ----------
+
+    seed : the seed to pass to np.random, mx.random and python's random.
+
+    To impose rng determinism, invoke e.g. as in:
+
+    with random_seed(1234):
+        ...
+
+    To impose rng non-determinism, invoke as in:
+
+    with random_seed():
+        ...
+
+    Upon conclusion of the block, the rng's are returned to
+    a state that is a function of their pre-block state, so
+    any prior non-determinism is preserved.
+
+    """
+
+    try:
+        next_seed = np.random.randint(0, np.iinfo(np.int32).max)
```

Comment:
Could you elaborate how this restores the pre-block state? To me this looks like it will always pick a new random seed after this block finishes. I rather expected that the old sequence generator is being restored. Imagine two sequences:
``1: 100 -> 101 -> 102 -> ...``
``2: 200 -> 201 -> 202 -> ...``

What I would expect is something along the lines of:
```
100
101
START: [random_seed(seq_2)]
200
201
202
END: [random_seed(seq_2)]
102
103
...
```

Reply:
You're asking about the "with random_seed(NNN)" construct (as opposed to the with_seed() decorator.  One issue faced here is that the mxnet rng state is not save-and-restorable, so we can't achieve exactly what you were expecting.  What's promised is this:
```
    Upon conclusion of the block, the rng's are returned to
    a state that is a function of their pre-block state, so
    any prior non-determinism is preserved.
```
So 'with random_seed(NNN)' forces determinism on the random values seen inside its block, without changing the determinism/non-determinism of the values seen outside the block.  Likewise, 'with random_seed()' (so no numeric arg) forces non-determinism on the random values seen inside its block, without changing the determinism/non-determinism of the values seen inside the block.

Let's take a use-case that explores this feature: a test that initially gets both random shapes and random data to populate tensors of those shapes.  Let's say the test is flaky and it's determined that it's best to work with a known set of shapes.  The test could be then coded as:
```
@with_seed()
def test_stuff():
    # seed 1234 produces a variety of shapes known to work with this test
    with random_seed(1234):
        shapes = <get 'random' shapes>
    # fill shapes with random data
    tensors = populate_tensors(shapes)
    <perform test>
```
Let's say now that the test is seen to fail sporadically, and  test seed 987654321 exposes the problem.  The methodology of this PR should still work:
```
MXNET_TEST_SEED=987654321 nosetests --verbose -s test_stuff_module.py:test_stuff
```
will repro the failure.

Reply:
Thanks for elaborating, this sounds very reasonable.

I just imagined a few nested blocks with random and hardcoded seeds and in my mind, your process seems to add up. None the less, could you please create a few unit tests to ensure that this behaviour works as expected and preserves a consistent sequence of number even if multiple blocks are nested, some seeds are hardcoded and so on? I think it's reasonable to validate our assumptions in case future changes get introduced.

Please also make sure to test the GPU context intensively, as we've experienced (known) issues on multi-GPU systems.